pkg/storegateway/bucket.go

```diff
@@ -615,102 +620,269 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
+	// Also, these need to be sent before we send IsEndOfSeriesStream=true.
+	if err := s.sendHintsAndStats(srv, resHints, stats); err != nil {
+		return err
+	}
+
+	// Send any remaining series and signal that there are no more series.
+	msg := &grpc.PreparedMsg{}
+	seriesBatch.IsEndOfSeriesStream = true
+	if err := msg.Encode(srv, storepb.NewStreamSeriesResponse(seriesBatch)); err != nil {
+		return status.Error(codes.Internal, errors.Wrap(err, "encode streaming series response").Error())
+	}
+	// Send the message.
+	if err := srv.SendMsg(msg); err != nil {
+		return status.Error(codes.Unknown, errors.Wrap(err, "send streaming series response").Error())
+	}
+
+	if seriesSet.Err() != nil {
+		return errors.Wrap(seriesSet.Err(), "expand series set")
+	}
+
+	return nil
+}
+
+func (s *BucketStore) sendSeriesChunks(
+	req *storepb.SeriesRequest,
+	srv storepb.Store_SeriesServer,
+	seriesSet storepb.SeriesSet,
+	stats *safeQueryStats,
+) (err error) {
+	var (
+		iterationBegin  = time.Now()
+		encodeDuration  = time.Duration(0)
+		sendDuration    = time.Duration(0)
+		seriesCount     int
+		chunksCount     int
+		streamingChunks = req.StreamingChunksBatchSize > 0 && !req.SkipChunks
+	)
 
-		err = nil
+	// Once the iteration is done we will update the stats.
+	defer stats.update(func(stats *queryStats) {
+		stats.mergedSeriesCount += seriesCount
+		stats.mergedChunksCount += chunksCount
+
+		// The time spent iterating over the series set is the
+		// actual time spent fetching series and chunks, encoding and sending them to the client.
+		// We split the timings to have a better view over how time is spent.
+		stats.streamingSeriesFetchSeriesAndChunksDuration += stats.streamingSeriesWaitBatchLoadedDuration
+		stats.streamingSeriesEncodeResponseDuration += encodeDuration
+		stats.streamingSeriesSendResponseDuration += sendDuration
+		stats.streamingSeriesOtherDuration += time.Duration(util_math.Max(0, int64(time.Since(iterationBegin)-
+			stats.streamingSeriesFetchSeriesAndChunksDuration-encodeDuration-sendDuration)))
 	})
 
-	if err != nil {
-		return
+	var batchSizeBytes int
+	var chunksBuffer []*storepb.StreamSeriesChunks
+	if streamingChunks {
+		chunksBuffer = make([]*storepb.StreamSeriesChunks, req.StreamingChunksBatchSize)
+		for i := range chunksBuffer {
+			chunksBuffer[i] = &storepb.StreamSeriesChunks{}
+		}
+	}
+	chunksBatch := &storepb.StreamSeriesChunksBatch{
+		Series: chunksBuffer[:0],
+	}
+	for seriesSet.Next() {
+		// IMPORTANT: do not retain the memory returned by seriesSet.At() beyond this loop cycle
+		// because the subsequent call to seriesSet.Next() may release it.
+		lset, chks := seriesSet.At()
+		seriesCount++
+		var response interface{}
+		if streamingChunks {
+			// We only need to stream chunks here because the series labels have already
+			// been sent above.
+			chunksBatch.Series = chunksBatch.Series[:len(chunksBatch.Series)+1]
+			last := chunksBatch.Series[len(chunksBatch.Series)-1]
+			last.Chunks = chks
```

Comment:
We have a race here. We should be making a copy here since it can be released by the iterator. I am on it.